staging/src/k8s.io/apiserver/pkg/storage/etcd3/featuresupportchecker/feature_support_checker.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package featuresupportchecker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+var (
+	// This is a error type that we used in checking RequestWatchProgress support
+	errNoSupportRequestWatchProgress = errors.New("requestWatchProgress is not supported for this cluster")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+	// featureSupportCache store result of if feature is supported or not.
+	featureSupportCache sync.Map
+)
+
+// FeatureSupportChecker to define Supports function which checks etcd with assigned feature and return boolean.
+type FeatureSupportChecker interface {
+	Supports(ctx context.Context, feature storage.Feature) (bool, error)
+}
+
+// featureSupportInfo holds information about feature support status and associated error.
+type featureSupportInfo struct {
+	Supported bool
+	Err       error
+}
+
+type DefaultFeatureSupportChecker struct {
+	client *clientv3.Client
+}
+
+// NewDefaultFeatureSupportChecker creates a DefaultFeatureSupportChecker with etcd3 client
+func NewDefaultFeatureSupportChecker(client *clientv3.Client) *DefaultFeatureSupportChecker {
+	return &DefaultFeatureSupportChecker{
+		client: client,
+	}
+}
+
+// Supports checks whether a specific feature is supported by etcd.
+// TODO: Extend this implementation to support feature-checking for the cacher.
+func (c *DefaultFeatureSupportChecker) Supports(ctx context.Context, feature storage.Feature) (bool, error) {
+	// Check if the enabled feature is already cached
+	if info, ok := featureSupportCache.Load(feature); ok {
```

Comment:
Cache should be per endpoint, not per feature.

Reply:
I agree about idea but we call `Supports` by feature not endpoint. Where do we want to use with endpoint?

Reply:
We don't care about other features, just return false.

Reply:
Ok let me rewrite this section real quick. The whole time i was thinking about this features and other features that might get implemented.

Reply:
Updated! 
PTAL. Thanks.